analytics_data_api/v0/tests/views/test_learners.py

```diff
@@ -404,9 +412,192 @@ def test_pagination(self):
+        """
+        self.create_learners([
+            {'username': 'user_1', 'course_id': self.course_id, 'segments': ['struggling', 'disengaging']},
+            {'username': 'user_2', 'course_id': self.course_id, 'segments': ['disengaging']}
+        ])
+        expected = self._get_full_engagement_ranges()
+        response = self._get(self.course_id)
+        self.assertEqual(response.status_code, 200)
+        expected['segments'] = {'disengaging': 2, 'struggling': 1, 'highly_engaged': 0, 'inactive': 0, 'unenrolled': 0}
+        self.assertDictContainsSubset(expected, json.loads(response.content))
+
+    @ddt.data(
+        [],
+        ['honor'],
+        ['verified'],
+        ['audit'],
+        ['nonexistent-enrollment-tracks-still-show-up'],
+        ['honor', 'verified', 'audit'],
+        ['honor', 'honor', 'verified', 'verified', 'audit', 'audit'],
+    )
+    def test_enrollment_modes(self, enrollment_modes):
+        self.create_learners([
+            {'username': 'user_{}'.format(i), 'course_id': self.course_id, 'enrollment_mode': enrollment_mode}
+            for i, enrollment_mode in enumerate(enrollment_modes)
+        ])
+        expected = self._get_full_engagement_ranges()
+        response = self._get(self.course_id)
+        self.assertEqual(response.status_code, 200)
+        expected_enrollment_modes = {}
+        for enrollment_mode, group in groupby(enrollment_modes):
+            # can't call 'len' directly on a group object
+            count = len([mode for mode in group])
+            expected_enrollment_modes[enrollment_mode] = count
+
+        expected['enrollment_modes'] = expected_enrollment_modes
+        self.assertDictContainsSubset(expected, json.loads(response.content))
+
+    @property
+    def empty_engagement_ranges(self):
+        """ Returns the engagement ranges where all fields are set to None. """
+        empty_engagement_ranges = {
+            'engagement_ranges': {
+                'date_range': {
+                    'start': None,
+                    'end': None
+                }
+            }
+        }
+        empty_range = {
+            range_type: [None, None] for range_type in ['below_average', 'average', 'above_average']
+        }
+        for metric in self.engagement_metrics:
+            empty_engagement_ranges['engagement_ranges'][metric] = copy.deepcopy(empty_range)
+        return empty_engagement_ranges
+
+    @property
+    def engagement_metrics(self):
+        """ Convenience method for getting the metric types. """
+        metrics = []
+        for entity_type in engagement_entity_types.AGGREGATE_TYPES:
+            for event in engagement_events.EVENTS[entity_type]:
+                metrics.append('{0}_{1}'.format(entity_type, event))
+        return metrics
+
+    # TODO: enable during https://openedx.atlassian.net/browse/AN-6319
+    # def test_cohorts(self):
+    #     pass
+
+    def test_no_engagement_ranges(self):
```

Comment:
I'm curious about this. The way we handled the 'no data' case with the timeline view was to return a 404 here. Then again, I think this makes sense, because we want to still be able to get segments and other data from this endpoint even if not all the data is available.